[ESP32S2] Support ESP32S2 platform

TD-er opened the pull request:
See also: https://github.com/letscontrolit/ESPEasy/issues/3578

State:
It compiles, and does crash very well and reliable.

TD-er committed:
[ESP32-S2] Add experimental support for ESP32-S2

TD-er committed:
[ESP32-S2] Further make ESP32-S2 compile (no WiFi yet)

TD-er committed:
[ESP32-S2] Add reset reasons

TD-er committed:
Merge branch 'mega' into feature/ESP32_S2_support

TD-er committed:
Merge branch 'mega' into feature/ESP32_S2_support

TD-er committed:
Merge branch 'mega' into feature/ESP32_S2_support

TD-er committed:
Merge branch 'mega' into feature/ESP32_S2_support

TD-er committed:
Merge branch 'mega' into feature/ESP32_S2_support

TD-er committed:
[ESP32-S2] Make it compile against esp32-2.0.0-pre

TD-er committed:
[ESP32-S2] Add missing includes

TD-er committed:
[ESP32-S2] Changed struct sizes for esp32-2.0.0-pre

TD-er commented:
See here for a test build: https://github.com/letscontrolit/ESPEasy/issues/3578#issuecomment-886932817

TD-er committed:
[ESP32-S2] Compute LogStruct size for new String size

Apparently the `String` object now is 4 bytes larger compared to before.

TD-er committed:
[ESP32-S2] Add GPIO definitions for ESP32-S2

TD-er committed:
ESPEasySerial to 2.0.6 to support ESP32-S2

TD-er commented:
@fmuntean If you build it yourself, with the commit included to support the GPIO-definitions of the ESP32-S2 I constantly get reboots, but as soon as I exclude that commit it works just fine.
You experience the same?

TD-er committed:
[PIO] Force ESPEasySerial to update to @ 2.0.6

TD-er committed:
[ESP32-S2] Fix build warnings + set ESPEasySerial lib to 2.0.6 for ESP32

fmuntean commented:
> @fmuntean If you build it yourself, with the commit included to support the GPIO-definitions of the ESP32-S2 I constantly get reboots, but as soon as I exclude that commit it works just fine.
> You experience the same?
@TD-er 
Yes, I get the same issue.
I got that this reboot at the begining as the I2C definition for ESP32 was IO21 and IO22 but for ESP32S2 the IO22 does not exists and trying to use that IO seems to trigger the reboot.
Maybe the same issue is here with the same or other IO.

TD-er committed:
[ESP32-S2] Add touch pins to ESP32Touch plugin + check valid GPIO I2C

tonhuisman commented on src/_P097_Esp32Touch.ino:
```diff
@@ -204,26 +207,46 @@ void P097_setEventParams(int pin, uint16_t threshold) {
       case 9: touchAttachInterrupt(T9, P097_got_T9, threshold); break;
+#ifdef ESP32S2
+/*
+      case 10: touchAttachInterrupt(T10, P097_got_T10, threshold); break;
```
Not enabled/activated yet?

TD-er replied:
Still trying to debug why some builds keep crashing when flashed.
It seems somewhat random whether a build is working or not.

TD-er committed:
[ESP32-S2] Use proper WiFi event ids for IDF > 3

TD-er committed:
[Cleanup] Check for validGpio using new function

Better to have a single function to check for valid GPIO pin instead of all kinds of checks like `!= -1` or `>= 0` or `> -1` etc.

TD-er committed:
[Cleanup] Include ESPEasyLimits.h where MAX_GPIO is used

TD-er commented:
If just checking out [[ESP32-S2] Use proper WiFi event ids for IDF > 3](https://github.com/letscontrolit/ESPEasy/pull/3579/commits/617d85522979453551ff561d173904c632d7d091) it wil compile into something that's just working (at least when built on Windows)
I also tried to build using the latest https://github.com/espressif/arduino-esp32/tree/2.0.0-rc1 but that one doesn't even allow to to link to a usable binary.

Since it is clearly a hit or miss without clear indication of what causes the crashes, I will now stop until the core library is in a more stable state.
It has taken me almost 20 hours again to get this working, so for now I have to leave it.

Maybe @Jason2866 does have some idea here, why changing seemingly unrelated code results in boot loops like these:

```
len:0x4b0
load:0x4004c000,len:0xa50
load:0x40050000,len:0x28e8
SHA-256 comparison failed:
Calculated: 2d04894f4d176b806e7e3761811f63f415223fc9d9980ca46ea6aacd6c079bd2
Expected: 1759f975513717b714ddbd483f4673fa88d9bd97b1991a82b3269740017349d2
Attempting to boot anyway...
entry 0x4004c190
ESP-ROM:esp32s2-rc4-20191025
Build:Oct 25 2019
rst:0x7 (TG0WDT_SYS_RST),boot:0x8 (SPI_FAST_FLASH_BOOT)
Saved PC:0x4002d735
SPIWP:0xee
mode:DOUT, clock div:1
load:0x3ffe6100,len:0x4b0
load:0x4004c000,len:0xa50
load:0x40050000,len:0x28e8
SHA-256 comparison failed:
Calculated: 2d04894f4d176b806e7e3761811f63f415223fc9d9980ca46ea6aacd6c079bd2
Expected: 1759f975513717b714ddbd483f4673fa88d9bd97b1991a82b3269740017349d2
Attempting to boot anyway...
entry 0x4004c190
ESP-ROM:esp32s2-rc4-20191025
Build:Oct 25 2019
rst:0x7 (TG0WDT_SYS_RST),boot:0x8 (SPI_FAST_FLASH_BOOT)
....
```

TD-er committed:
Merge branch 'mega' into feature/ESP32_S2_support

Jason2866 commented:
Linking does fail when not using esptool.py v.3.1 !
Flash mode qio does NOT work https://github.com/letscontrolit/ESPEasy/pull/3579/files#diff-86bfbbaa822cbc1b9f32448e1e6e88914aa63ae090e25f5db885b6bf1215a346R11
We use our own boards definitions. Some have errors. Doing this has simpified platformio setup too. Our env for Tasmota. 
```
;*** EXPERIMENTAL Tasmota version for ESP32-S2
[env:tasmota32s2]
extends                     = env:tasmota32_base
board                       = esp32s2
platform                    = https://github.com/platformio/platform-espressif32.git#feature/arduino-idf-master
platform_packages           = framework-arduinoespressif32 @ https://github.com/Jason2866/esp32-arduino-lib-builder/releases/download/375/framework-arduinoespressif32-master-32a9e33b1.tar.gz
                              platformio/tool-mklittlefs @ ~1.203.200522
build_unflags               = ${esp32_defaults.build_unflags}
build_flags                 = ${esp32_defaults.build_flags} -DFIRMWARE_LITE
lib_extra_dirs              = lib/libesp32
                              lib/lib_basic
lib_ignore                  =
    NimBLE-Arduino
    Micro-RTSP
    ESP32-HomeKit
```
You can use the Arduino32 repo branch as source too. We have changed some sdkconfig settings in IDF 4.4 for Tasmota to get a smaller firmware.
Our boards definitions are here https://github.com/arendst/Tasmota/tree/development/boards

Jason2866 commented:
Not saying Tasmota does run on the S2. We have to search why LittleFS is not working. LittleFS is the standard Filesystem for all Tasmota variants. S2 is the only variant which we have problems with. S2 is not our focus. The ESP32-C3 is the one we did invest time. Surprisingly it runs already very stable :-)

TD-er commented:
OK, thanks.
I will order a few of those ESP32-C3 then.
And I will test to see if I can get it to work with your hints too.
It really feels like "hit-or-miss" with each ESP32-S2 build.

Jason2866 commented:
There are C3 pinout compatible ones to the ESP-12F modules. Great for replacing Tuya modules (with non DIY friendly MCUs inside). But i think thats not ESPEasy main focus...

TD-er commented:
Good to know they are compatible with those ESP12F modules.
It may not be the main focus, but I do have a lot of boards laying around (still not soldered) which have pads for the ESP12-F modules.
So who knows :)

TD-er commented:
Hmm with the PIO config you gave I could build it and upload, but the unit does seem to remain stuck with messages like these:
```
ESP-ROM:esp32s2-rc4-20191025
Build:Oct 25 2019
rst:0x1 (POWERON),boot:0x8 (SPI_FAST_FLASH_BOOT)
SPIWP:0xee
mode:DOUT, clock div:1
load:0x3ffe6100,len:0x7c
load:0x4004c000,len:0x818
load:0x40050000,len:0x22bc
entry 0x4004c168
��������������������߾����������IA@�$�HP�6�� ������$C��2A�_ �ESP-ROM:esp32s2-rc4-20191025
Build:Oct 25 2019
rst:0xc (RTC_SW_CPU_RST),boot:0x8 (SPI_FAST_FLASH_BOOT)
Saved PC:0x40026700
SPIWP:0xee
mode:DOUT, clock div:1
load:0x3ffe6100,len:0x7c
load:0x4004c000,len:0x818
load:0x40050000,len:0x22bc
entry 0x4004c168
��������������������������������4�!@�0� P�I��C2"    ��d$�ESP-ROM:esp32s2-rc4-20191025
```

So have to test a bit more.
But it just feels like the ESP32-S2 support is far from finished.
Not at all "RC" worthy.
Like I said, with previous "april" version of the library it would run just fine, until you change something seemingly unrelated and the node will boot-loop forever.

TD-er committed:
Merge branch 'mega' into feature/ESP32_S2_support

TD-er commented:
@mcspr I just rebased this PR to the latest code in the mega branch.
Now it does seem to be working again, where I was not able to build a version that finished booting as it was crashing like in the previous comment.
Not sure if the `-fno-strict-aliasing` f;ag did actually fix it or not, but at least now it does produce a working build again.

TD-er commented:
OK, I stand corrected...
It looks like this build flag is _not_ used on `custom_ESP32s2_4M316k`.
So I added it and was seeing crashes again.
Completely cleaned all caches, downloaded packages and libs and built again -> crashes.
Removed the build flag, clean build -> Working again.

So for this specific build issue, the no strict aliasing is not the magical fix. :(

Jason2866 commented:
LittleFS does not work with S2. I opened a issue https://github.com/espressif/arduino-esp32/issues/5550
s-hadinger (the author of the zigbee part, lvgl and berry implementation for Tasmota) did a few test and stopped. He is thinking:
> I tested and it's still broken on my side. Reading the Flash returns garbage, so I suspect a wrongly initialized Flash SPI configuration (which is weird because code is still being pulled from SPI Flash)

TD-er commented:
OK, that could also explain strange reboot loops when making no related changes in the code.
Sounds a bit like a lost cause, the ESP32s2. It is really hit-or-miss with each build for it.

Jason2866 commented:
A little progress. Tasmota does boot. Look at the flash size which Tasmota gets. Now i am sure SPI flash size handling is screwed up in Arduino.
![Screenshot_20210818-154743](https://user-images.githubusercontent.com/24528715/129913140-b1959244-b7af-4df6-b0b8-86151b10d4ba.png)
Boot log, LittleFS does fail because flash writing is faulty:
```
ESP-ROM:esp32s2-rc4-20191025
Build:Oct 25 2019
rst:0x1 (POWERON),boot:0x8 (SPI_FAST_FLASH_BOOT)
SPIWP:0xee
mode:DOUT, clock div:1
load:0x3ffe6100,len:0x7c
load:0x4004c000,len:0x810
load:0x40050000,len:0x23e4
entry 0x4004c164
ßHø00:00:00.001 HDW: ESP32-S2
./components/esp_littlefs/src/littlefs/lfs.c:998:error: Corrupted dir pair at {0x0, 0x1}
E (451) esp_littlefs: mount failed,  (-84)
E (452) esp_littlefs: Failed to initialize LittleFS
./components/esp_littlefs/src/littlefs/lfs.c:1685:debug: Bad block at 0x0
./components/esp_littlefs/src/littlefs/lfs.c:1691:warn: Superblock 0x0 has become unwritable
E (1730) esp_littlefs: Failed to format filesystem
00:00:00.287 CFG: Loaded from Nvm, Count 1321859368
00:00:00.291 CFG: Use defaults
00:00:00.384 QPC: Reset
00:00:00.399 BRY: Berry initialized, RAM used=1779
00:00:00.401 BRY: no 'preinit.be'
00:00:00.404 Project tasmota - Tasmota Version 9.5.0.5(tasmota)-2_0_0_rc(2021-08-16T18:18:35)
00:00:00.411 BRY: no 'autoexec.be'
00:00:00.710 WIF: WifiManager active for 3 minutes
00:00:01.107 HTP: Web server active on tasmota-0C0A3C-2620 with IP address 192.168.4.1
00:00:07.115 QPC: Reset
E (11937) esp_image: image at 0x10000 has invalid magic byte (nothing flashed here?)
```

TD-er commented:
I'm using SPIFFS:
![image](https://user-images.githubusercontent.com/3751318/129938564-e5bab06a-3986-4190-9d62-22b535e758c5.png)

But maybe you can also test to change only a bit of code (e.g. switching languages in Tasmota build) so you know the objects will be linked at different positions and try to see if the unit boots.
It really is a hit-or-miss with this module.

Also the WiFi does reply a bit strange, as it may end up in limbo state when it needs to reconnect.
You cannot set the WiFi mode anymore and if that happens you need to reboot (power cycle), or at least I have not yet managed to get the WiFi to behave well after it got in such a state.

Jason2866 commented:
@TD-er  Fully working Tasmota running on a ESP32-S2 (every build)  :-)
```
ESP-ROM:esp32s2-rc4-20191025
Build:Oct 25 2019
rst:0x1 (POWERON),boot:0x8 (SPI_FAST_FLASH_BOOT)
SPIWP:0xee
mode:DOUT, clock div:1
load:0x3ffe6100,len:0x7c
load:0x4004c000,len:0x7e4
load:0x40050000,len:0x2598
entry 0x4004c15c

00:00:00.001 HDW: ESP32-S2 (PSRAM)
00:00:00.006 UFS: FlashFS mounted with 304 kB free
00:00:00.018 CFG: Loaded from File, Count 4
E (10051) tsens: Do not configure the temp sensor when it's running!
```
![image](https://user-images.githubusercontent.com/24528715/131889560-2251145e-d31f-42a2-ba95-3ab5e08d6353.png)
the env
```
;*** EXPERIMENTAL Tasmota version for ESP32-S2 (Tasmota Arduino ESP32 v.2.0.0) 
[env:tasmota32s2]
extends                 = env:tasmota32_base
board                   = esp32s2
upload_port             = COM5
platform                = https://github.com/platformio/platform-espressif32.git#feature/arduino-idf-master
platform_packages       = framework-arduinoespressif32 @ https://github.com/Jason2866/esp32-arduino-lib-builder/releases/download/424/framework-arduinoespressif32-master-cdcf92440.tar.gz
                          platformio/tool-mklittlefs @ ~1.203.200522
build_unflags           = ${env:tasmota32_base.build_unflags}
                          -Wswitch-unreachable
                          -Wstringop-overflow
                          -Wincompatible-pointer-types
build_flags             = ${env:tasmota32_base.build_flags}
                          -Wno-switch-unreachable
                          -Wno-stringop-overflow
```

TD-er commented:
Thanks, will try later this evening.

Did you have issues before ESP32 v2.0.0 (release of 2 days ago) ?

And another one for you to test, I have seen that my ESPEasy units did not reconnect to WiFI if connection was lost.
Not looked very much into it, but it seems like I may not get the right disconnect event.
Maybe I have to take a closer look to the (new) event nrs, maybe I have missed some?

Jason2866 commented:
Yes, we had issues with the official 2.0 version. The sdkconfig settings choosen in IDF for the builded Arduino libs are not optimal for Tasmota needs. For the S2 i never got LittleFS working.
With the setup we have choosen it does work. Example working S2 env
```
[env:esp32-s2-saola-1]
platform = https://github.com/platformio/platform-espressif32.git#feature/arduino-idf-master
platform_packages = framework-arduinoespressif32 @ https://github.com/tasmota/arduino-esp32/releases/download/2.0.0/framework-arduinoespressif32-release_IDF4.4.tar.gz
board = esp32-s2-saola-1
framework = arduino
``` 
No reconnect issues, so far

TD-er commented:
OK, still have to test the S2 board with it.
But last night I also tested the latest platform you described with ESP-now and there seem to be some fixes regarding the interoperability between ESP32 and ESP8266.
Where the (PlatformIO) revision 3.x was hardly to not at all able to communicate between platforms, this newer set does seem to be almost usable when using ESP-now between platforms.
At least now there are no longer SDK debug logs sent to Serial0 if you hit some obscure errors like running out of peer slots.

Will try to build it -hopefully- this evening on the S2 board to see if that will result in more reliable builds.

TD-er committed:
Merge branch 'mega' into feature/ESP32_S2_support

TD-er committed:
[ESP32-S2] Switch platform packages for ESP32s2 to improve build success

As suggested here: https://github.com/letscontrolit/ESPEasy/pull/3579#issuecomment-912620988 by @Jason2866

TD-er commented:
OK, finally got to make a build for it on my PC and with these suggestions it now is successful building, while it wasn't before.

So for now, this seems a success. Thanks @Jason2866

TD-er commented:
@Jason2866 I've been testing a bit with my current build for the ESP32-S2 and there is still a lot to do.
One of the things is that simply scanning the I2C bus triggers a reboot.
Did you find something similar?
Still have to hook up the debugger/stackdump to it, but so far the reboot reason is simply a "Manual Reboot". (no idea what it indicates here.....)

TD-er committed:
[ESP32-s2] Suppress SDK info/debug logs on Serial0

TD-er committed:
[ESP32-s2] Fix crash when scanning I2C bus

See: https://github.com/espressif/arduino-esp32/issues/5636

TD-er commented:
@Jason2866 I got the crash at I2C scan fixed and now also the WiFI may reconnect.
Except the WiFi is terribly slow after a reconnect.
Even pings timeout, or reply with a response time of upto 3 seconds.
Also the web interface replies really slow then.

Anything sounding familiar?

TD-er committed:
[WiFi] Only set WiFi initialised when connected & got IP

TD-er committed:
[WiFi] Clear IP config on disconnect

Jason2866 commented:
No, the S2 Tasmota build has no wifi issues.

TD-er committed:
Merge branch 'mega' into feature/ESP32_S2_support

TD-er committed:
Merge branch 'mega' into feature/ESP32_S2_support

TD-er committed:
[ESP32-S2] Use latest ESP32 core 2.0.0 + IDF4.4 + I2C fix

See: https://github.com/espressif/arduino-esp32/issues/5636

TD-er committed:
[ESP32-S2] Mark ESP32-s2 and ESP32-c3 as ESPEasy platforms

TD-er committed:
[PSRAM] Add preliminary support for PSRAM

This will later be used just like the 2nd heap on core 3.0.0 for ESP8266.
For now just getting the info and it will be used as extra for heap allocation when needed.

TD-er commented:
OK, this now does work quite stable already.
Big thanks to @Jason2866  for providing the toolchain for this.

Still need to do a lot of cleanup regarding check for correct reset reasons, WiFi events and some actual testing using sensors.
But I already feel like it is ready to be merged, so I can get rid of some long standing parallel developing branches which in total have taken quite some time to develop.

TD-er committed:
Merge branch 'mega' into feature/ESP32_S2_support

TD-er committed:
[ESP32 IDF4.x] Fix build issue with ESP32 IDF 4.x

TD-er committed:
[P109 ThermOLED] Fix array bound error

TD-er committed:
Fix build error on limited build size builds

Jason2866 commented:
@TD-er  This framework does check for PSRAM at boot https://github.com/Jason2866/esp32-arduino-lib-builder/blob/master/sdkconfig.esp32#L509-L530 and does NOT abort if none found.
We use this for autodetecting PSRAM and use when found. The program has to take care to disable PSRAM for ESP32 revisions which needs the PSRAM fix files linked. If the fix files are linked for this revisions PSRAM will be enabled too. Doing this way we avoid unnecessary linking of this files.
ESP32-S2 has NO PSRAM issues!

TD-er commented:
Yep, I  have read something about it in the IDF documentation, where those issues were mainly in older than rev. 3 silicon.
As you may have seen in my PR regarding the ESP8266 core 3.0.0 support, I was working on stuffing typical "runtime" data like logs, controller queue, WiFi scans, etc in the 2nd heap.
I plan to do the same with the PSRAM, so I don't need to think about what data to store where when loading/saving on flash.

For now that's the main purpose for me to start using the PSRAM, so I do not (yet) plan on using it to increase the main heap allocation pool.

I am still thinking about whether or not I plan on adding it to regular ESP32 builds, mainly due to the "PSRAM issues" you also mentioned and I don't want to double the number of builds with PSRAM enabled/disabled.
So it will probably be like "enabled on boards without issues" and "possible to enable on custom builds".

I still have other ideas for future features regarding data processing, which may benefit from large amounts of RAM (at least few 100 k) to be able to create "derived" sensors, by combining sensors.
For this I also may need a very accurate time base, so that's why I have spent a lot of time on synchronizing it among nodes for the ESP-now layer I'm working on. (which you've probably also looked at :) )

Thanks for the links, will look into it.
N.B. I also read about some "Allow to abort" compile flag for silicon that does have issues when no PSRAM is found.

Jason2866 commented:
We do not wanted extra builds for Tasmota when using PSRAM too. So we have taken care that PSRAM is only enabled and used for ESP32 revisions which do not need the fix libs. There is a log message in Tasmota in this case PSRAM found but not used since the fix libs are needed and not there. So the user knows he can use when he does a custom build. Tasmota uses PSRAM for the Berry scripting language (something between LUA and MicroPython) and the universal display LVGL driver.

TD-er committed:
[Cleanup] Clean up of double includes + pio.ini description

TD-er committed:
[NeoPixel] Update to latest Adafruit Neopixel library

To fix support for NeoPixel in ESP32/ESP32-s2

TD-er committed:
[Docs] Add some description of ESP82xx/ESP32/ESP32-S2

Jason2866 commented:
@TD-er did a Platformio version of the released Arduino Esp32 core 2.0.1 for Tasmota.
Many fixes since the version you actual use. Since yesterday we build all Tasmota ESP32xx with.
```
platform                    = https://github.com/platformio/platform-espressif32.git#feature/arduino-idf-master
platform_packages           = framework-arduinoespressif32 @ https://github.com/tasmota/arduino-esp32/releases/download/2.0.1/framework-arduinoespressif32-release_IDF4.4.tar.gz
                              platformio/tool-esptoolpy @ https://github.com/tasmota/esptool/releases/download/v3.2/esptool-v3.2.zip
```

TD-er commented:
Thanks, will check it immediately as I am having some issues which are probably already long fixed regarding TLS.

Did you find any differences in WiFi responsiveness with the 2.0.x for ESP32?
That's one of the reasons I was still using an older one, because I noticed some packet loss in some setups using ESP32 (not -S2) which was working just fine with the older platforms.

TD-er commented:
@Jason2866 It seems like the build size is quite a lot smaller.
In my test here (not yet ran it) the same build went from 1776 kB to 1370 kB

Jason2866 commented:
Yes, removed assertions. You will get more free RAM too

TD-er commented:
Not seeing the extra RAM, or at least it isn't that much.

But the ping issues are back, just as in my previous tests.

```
C:\Users\gijsn>ping 192.168.10.174 -n 20

Pinging 192.168.10.174 with 32 bytes of data:
Request timed out.
Reply from 192.168.10.174: bytes=32 time=3449ms TTL=255
Reply from 192.168.10.174: bytes=32 time=10ms TTL=255
Reply from 192.168.10.174: bytes=32 time=1102ms TTL=255
Reply from 192.168.10.174: bytes=32 time=3ms TTL=255
Reply from 192.168.10.174: bytes=32 time=673ms TTL=255
Request timed out.
Reply from 192.168.10.174: bytes=32 time=134ms TTL=255
Reply from 192.168.10.174: bytes=32 time=209ms TTL=255
Request timed out.
Request timed out.
Request timed out.
Request timed out.
Reply from 192.168.10.174: bytes=32 time=2741ms TTL=255
Reply from 192.168.10.174: bytes=32 time=15ms TTL=255
Request timed out.
Reply from 192.168.10.174: bytes=32 time=151ms TTL=255
Reply from 192.168.10.174: bytes=32 time=447ms TTL=255
Request timed out.
Request timed out.

Ping statistics for 192.168.10.174:
    Packets: Sent = 20, Received = 11, Lost = 9 (45% loss),
Approximate round trip times in milli-seconds:
    Minimum = 3ms, Maximum = 3449ms, Average = 812ms
```

TD-er commented:
And with not changing the TX power and continuously sending Gratuituous ARP:

```
C:\Users\gijsn>ping 192.168.10.174 -n 20

Pinging 192.168.10.174 with 32 bytes of data:
Reply from 192.168.10.174: bytes=32 time=235ms TTL=255
Reply from 192.168.10.174: bytes=32 time=2408ms TTL=255
Reply from 192.168.10.174: bytes=32 time=4ms TTL=255
Reply from 192.168.10.174: bytes=32 time=2764ms TTL=255
Reply from 192.168.10.174: bytes=32 time=4ms TTL=255
Reply from 192.168.10.174: bytes=32 time=209ms TTL=255
Request timed out.
Reply from 192.168.10.174: bytes=32 time=3339ms TTL=255
Reply from 192.168.10.174: bytes=32 time=4ms TTL=255
Reply from 192.168.10.174: bytes=32 time=5ms TTL=255
Request timed out.
Reply from 192.168.10.174: bytes=32 time=3343ms TTL=255
Reply from 192.168.10.174: bytes=32 time=8ms TTL=255
Request timed out.
Reply from 192.168.10.174: bytes=32 time=350ms TTL=255
Reply from 192.168.10.174: bytes=32 time=3581ms TTL=255
Reply from 192.168.10.174: bytes=32 time=4ms TTL=255
Request timed out.
Request timed out.
Reply from 192.168.10.174: bytes=32 time=2935ms TTL=255

Ping statistics for 192.168.10.174:
    Packets: Sent = 20, Received = 15, Lost = 5 (25% loss),
Approximate round trip times in milli-seconds:
    Minimum = 4ms, Maximum = 3581ms, Average = 1279ms
```
Less packet loss, but that's not really true to the responsiveness as the ping replies take several seconds every N pings.

Jason2866 commented:
Mhh, with Tasmota I get this
```
Last login: Wed Nov 10 18:34:30 on ttys000
hans@MacBook-Air-von-Johann ~ % ping 192.168.2.166
PING 192.168.2.166 (192.168.2.166): 56 data bytes
64 bytes from 192.168.2.166: icmp_seq=0 ttl=255 time=8.084 ms
64 bytes from 192.168.2.166: icmp_seq=1 ttl=255 time=5.758 ms
64 bytes from 192.168.2.166: icmp_seq=2 ttl=255 time=12.883 ms
64 bytes from 192.168.2.166: icmp_seq=3 ttl=255 time=5.238 ms
64 bytes from 192.168.2.166: icmp_seq=4 ttl=255 time=7.624 ms
64 bytes from 192.168.2.166: icmp_seq=5 ttl=255 time=5.541 ms
64 bytes from 192.168.2.166: icmp_seq=6 ttl=255 time=7.050 ms
64 bytes from 192.168.2.166: icmp_seq=7 ttl=255 time=11.394 ms
64 bytes from 192.168.2.166: icmp_seq=8 ttl=255 time=8.588 ms
64 bytes from 192.168.2.166: icmp_seq=9 ttl=255 time=7.519 ms
64 bytes from 192.168.2.166: icmp_seq=10 ttl=255 time=5.415 ms
64 bytes from 192.168.2.166: icmp_seq=11 ttl=255 time=5.636 ms
64 bytes from 192.168.2.166: icmp_seq=12 ttl=255 time=7.297 ms
64 bytes from 192.168.2.166: icmp_seq=13 ttl=255 time=9.980 ms
64 bytes from 192.168.2.166: icmp_seq=14 ttl=255 time=5.157 ms
64 bytes from 192.168.2.166: icmp_seq=15 ttl=255 time=3.929 ms
64 bytes from 192.168.2.166: icmp_seq=16 ttl=255 time=5.430 ms
64 bytes from 192.168.2.166: icmp_seq=17 ttl=255 time=5.200 ms
64 bytes from 192.168.2.166: icmp_seq=18 ttl=255 time=10.041 ms
64 bytes from 192.168.2.166: icmp_seq=19 ttl=255 time=4.535 ms
64 bytes from 192.168.2.166: icmp_seq=20 ttl=255 time=7.289 ms
64 bytes from 192.168.2.166: icmp_seq=21 ttl=255 time=4.517 ms
64 bytes from 192.168.2.166: icmp_seq=22 ttl=255 time=5.898 ms
```

TD-er commented:
That's how it should be...
Problem is that this isn't happening with all ESP32 boards I have, so that may make it -again- some debugging session to see what is now the problem with the black box called ESP-wifi.

TD-er commented:
OK, spent lots of hours (> 10, so missing some sleep...) on this to see what is happening here.

So far I found that:
- hostname should be set on ESP32 _before_ calling `WiFi.mode`
- (re)connect without WiFi scan does fix the problem.

It seems the IDF 4.4 may keep the channel set to the last one used in a scan.
So performing async scan is a big no-no and/or you should reconnect after performing a scan.

May need a bit more to look into, but the ping times are 2 - 10 msec for my test setup now, where they were really abominable before (average 900 msec when running for hours)

Jason2866 commented:
Oh, sorry that i gave you "something" that was a sleep thief ;-)
Great finding.

TD-er commented:
Well I'm still trying to wrap my head around it to what actually causes these issues as it is for sure more complex than just not performing a WiFi scan.
So far it is working fine, if I let it connect, wait for a while and then let it disconnect + reconnect.
The reason it does not perform a scan then (in ESPEasy) is that it still considers the last scan as usable (not expired) and simply tries to connect to the same as before.
The WiFi is turned off in between, so it is essentially the same as on a cold boot.

However, if I don't call the ESPEasy command `wifidisconnect`, but `wifimode off`, it does result in a bad connection. (or not a usable connection at all)
So it seems like there is something really timing critical here and that's what I really don't like as it may need different timings on other nodes.

Jason2866 commented:
Just a shoot in the dark. Maybe the behavior is related to the ESP32 chip revision?
Similar to the PSRAM bug for v1 and v2. Maybe the revision have some more bugs which are solved in later revisions

TD-er commented:
As far as I can tell, it is now working on the problematic unit(s) too.
Could be it is a silicon bug, or maybe the tolerances for WiFi are made slightly more tight in IDF 4.4

The main issue is that you should not continue with an existing connection as soon as a WiFi scan completes (doesn't matter if it is synchronous or async).
And you should not reconnect too quickly after a disconnect.

Also it seems the ESP32 Arduino code base now has more .cpp files, as I ran into the imfamous build issues again where the linker fails due to max. command line length in Windows.
So I had to concatenate even more .cpp files during build to make sure I have less .o files to link to stay under that limit of 32768 characters on the linker command line.
